File: perf/benchmarks/apps/dijkstra/dijkstra.cpp
// dijkstra — dense-graph SSSP application benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <vector>

constexpr int INF = 1 << 30;

int main(int argc, char** argv) {
    int v = (argc > 1) ? std::atoi(argv[1]) : 800;
    int64_t seed = 1;
    std::vector<int> g(static_cast<size_t>(v) * v);
    for (int i = 0; i < v; ++i) {
        for (int j = 0; j < v; ++j) {
            seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
            g[i * v + j] = 1 + static_cast<int>((seed >> 8) % 100);
        }
        g[i * v + i] = 0;
    }

    auto t0 = std::chrono::steady_clock::now();
    std::vector<int> dist(v, INF);
    std::vector<uint8_t> visited(v, 0);
    dist[0] = 0;
    for (int s = 0; s < v; ++s) {
        int u = -1, best = INF;
        for (int k = 0; k < v; ++k) {
            if (!visited[k] && dist[k] < best) { best = dist[k]; u = k; }
        }
        if (u == -1) break;
        visited[u] = 1;
        int du = dist[u];
        const int* row = &g[u * v];
        for (int k = 0; k < v; ++k) {
            int nd = du + row[k];
            if (nd < dist[k]) dist[k] = nd;
        }
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    int64_t cs = 0;
    for (int d : dist) cs ^= d;
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/apps/dijkstra/dijkstra.py
# dijkstra — dense-graph SSSP application benchmark.
#
# Build a V*V adjacency matrix with integer weights from an LCG, then run
# Dijkstra's algorithm with an O(V^2) array-based priority. Returns the
# XOR-fold of the shortest distances.
# Argument: V (default 800).

import sys
import time

INF = 1 << 30


def main() -> None:
    v = int(sys.argv[1]) if len(sys.argv) > 1 else 800
    seed = 1
    g = [[0] * v for _ in range(v)]
    for i in range(v):
        for j in range(v):
            seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
            g[i][j] = 1 + ((seed >> 8) % 100)
        g[i][i] = 0

    t0 = time.perf_counter_ns()
    dist = [INF] * v
    visited = [False] * v
    dist[0] = 0
    for _ in range(v):
        u = -1
        best = INF
        for k in range(v):
            if not visited[k] and dist[k] < best:
                best = dist[k]
                u = k
        if u == -1:
            break
        visited[u] = True
        row = g[u]
        du = dist[u]
        for k in range(v):
            nd = du + row[k]
            if nd < dist[k]:
                dist[k] = nd
    elapsed_ns = time.perf_counter_ns() - t0

    cs = 0
    for d in dist:
        cs ^= d
    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
